### PR TITLE
CLDIDE-2638 Save snapshot in onprem version

### PR DESCRIPTION
**Pull request by skabashnyuk**

### CODE REVIEW

File: plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceKey.java [hunk 1]
```diff
@@ -10,57 +10,48 @@
...
 
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.Map;
+import org.eclipse.che.api.machine.server.impl.InstanceKeyImpl;
+import org.eclipse.che.api.machine.server.spi.InstanceKey;
 
 /**
- * Set of keys that identifies docker image properties
+ * Set of helper method that identifies docker image properties
```
Comment (garagatyi) [inline plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceKey.java]: Set of helper method -> Add set of helper methods

File: plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceKey.java [hunk 2]
```diff
@@ -10,57 +10,48 @@
...
+        super(ImmutableMap.of(REPOSITORY, repository, TAG, tag, ID, id, REGISTRY, registry));
+    }
+
+    public DockerInstanceKey(String repository, String id, String registry) {
+        super(ImmutableMap.of(REPOSITORY, repository, ID, id, REGISTRY, registry));
     }
 
     public String getRepository() {
-        return fields.get(REPOSITORY);
+        return getFields().get(REPOSITORY);
```
Comment (garagatyi) [inline plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceKey.java]: You create instance of immutable map on each call of this method. See my comment here https://github.com/codenvy/che-core/pull/575/files#r48469281

File: plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceKey.java [hunk 3]
```diff
@@ -10,57 +10,48 @@
...
+        super(key);
     }
 
     public DockerInstanceKey(String repository, String tag, String id, String registry) {
-        fields = new HashMap<>(4);
-        fields.put(REPOSITORY, repository);
-        fields.put(TAG, tag);
-        fields.put(ID, id);
-        fields.put(REGISTRY, registry);
+        super(ImmutableMap.of(REPOSITORY, repository, TAG, tag, ID, id, REGISTRY, registry));
```
Comment (garagatyi) [inline plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceKey.java]: I think you can move creation of immutable map to parent class.

File: plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceKey.java (hunk 2 above)
Comment (skabashnyuk) [inline plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceKey.java]: done

File: plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstance.java [hunk 4]
```diff
@@ -186,10 +186,12 @@ public InstanceKey saveToSnapshot(String owner) throws MachineException {
                 comment = comment + " by " + owner;
             }
             // !! We SHOULD NOT pause container before commit because all execs will fail
-            final String imageId = docker.commit(container, repository, null, comment, owner);
             // to push image to private registry it should be tagged with registry in repo name
             // https://docs.docker.com/reference/api/docker_remote_api_v1.16/#push-an-image-on-the-registry
-            docker.tag(imageId, registry + "/" + repository, null);
+            docker.commit(container, registry + "/" + repository, null, comment, owner);
+            //TODO fix this workaround. Docker image is not visible after commit when using swarm
+            Thread.sleep(2000);
```
Comment (garagatyi) [inline plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstance.java]: Please create issue for investigating that. Assign to me

**Comment by garagatyi** [comment]
ok

**Pull request merged**